src/feature/hs/hs_circuit.c

```diff
@@ -1197,20 +1215,17 @@ hs_circ_send_establish_rendezvous(origin_circuit_t *circ)
   return -1;
 }
 
-/** We are about to close or free this <b>circ</b>. Clean it up from any
- * related HS data structures. This function can be called multiple times
- * safely for the same circuit. */
+/** We are about to close this <b>circ</b>. Clean it up from any related HS
+ * data structures. This function can be called multiple times safely for the
+ * same circuit. */
 void
-hs_circ_cleanup(circuit_t *circ)
+hs_circ_cleanup_on_close(circuit_t *circ)
 {
   tor_assert(circ);
 
-  /* If it's a service-side intro circ, notify the HS subsystem for the intro
-   * point circuit closing so it can be dealt with cleanly. */
-  if (circ->purpose == CIRCUIT_PURPOSE_S_ESTABLISH_INTRO ||
-      circ->purpose == CIRCUIT_PURPOSE_S_INTRO) {
-    hs_service_intro_circ_has_closed(TO_ORIGIN_CIRCUIT(circ));
-  }
+  /* NOTE: Following code must remain light and fast. The circuit close
+   * function has to remain very fast since it is called as part of multiple
+   * main loop event. */
 
   /* Clear HS circuitmap token for this circ (if any). Very important to be
    * done after the HS subsystem has been notified of the close else the
```

Comment:
I'm a bit confused on what this comment means now. What does "important to be done after the HS subsystem has been notified of the close" in this context? I thought that now this function is responsible for notifying the HS subsystem of the close, so what is important to be done before the circuitmap clear?

Reply:
As a side note here, since we are being super nice and clean and making nice functions for our cleanup, I would really like a single high-level comment (on top of our cleanup functions) that details exactly our high-level cleanup strategy. We  are nicely detailing what each function should do, but it would be nice to have centralized what we want to have happened after all these functions are called.

Reply:
Good idea. See fixup commit: f50e95bb58